File: 2816.double-a-number-represented-as-a-linked-list.py
#
# @lc app=leetcode id=2816 lang=python3
#
# [2816] Double a Number Represented as a Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # If the head is greater than 4, doubling it will result in a carry
        # that requires a new node at the front.
        if head.val > 4:
            head = ListNode(0, head)
        
        current = head
        while current:
            # Double the current value and take modulo 10
            current.val = (current.val * 2) % 10
            
            # Check if the next node will generate a carry
            # A carry is generated if the next node's value is > 4 (since 5*2=10)
            if current.next and current.next.val > 4:
                current.val += 1
            
            current = current.next
            
        return head
# @lc code=end

File: 2816.double-a-number-represented-as-a-linked-list.cpp
#
# @lc app=leetcode id=2816 lang=cpp
#
# [2816] Double a Number Represented as a Linked List
#
# @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int doubleHelper(ListNode* node) {
        if (node == nullptr) return 0;
        
        int carry = doubleHelper(node->next);
        int val = node->val * 2 + carry;
        node->val = val % 10;
        return val / 10;
    }
    
    ListNode* doubleIt(ListNode* head) {
        int carry = doubleHelper(head);
        if (carry > 0) {
            ListNode* newHead = new ListNode(carry);
            newHead->next = head;
            head = newHead;
        }
        return head;
    }
};
# @lc code=end